tree-sitter/src/TreeSitter/Node.hs

```diff
@@ -72,73 +81,86 @@ instance Storable Node where
                            <*> peekStruct
                            <*> peekStruct
                            <*> peekStruct
-                           <*> peekStruct
-  poke ptr (Node n t s sp ep sb eb c nn ne) = flip evalStruct ptr $ do
+  {-# INLINE peek #-}
+  poke ptr (Node n t s ep eb c fn nn ne) = flip evalStruct ptr $ do
     pokeStruct n
     pokeStruct t
     pokeStruct s
-    pokeStruct sp
     pokeStruct ep
-    pokeStruct sb
     pokeStruct eb
     pokeStruct c
+    pokeStruct fn
     pokeStruct nn
     pokeStruct ne
+  {-# INLINE poke #-}
 
 instance Storable TSPoint where
   alignment _ = alignment (0 :: Int32)
+  {-# INLINE alignment #-}
   sizeOf _ = 8
+  {-# INLINE sizeOf #-}
   peek = evalStruct $ TSPoint <$> peekStruct
                               <*> peekStruct
+  {-# INLINE peek #-}
   poke ptr (TSPoint r c) = flip evalStruct ptr $ do
     pokeStruct r
     pokeStruct c
+  {-# INLINE poke #-}
 
 instance Storable TSNode where
   alignment _ = alignment (nullPtr :: Ptr ())
+  {-# INLINE alignment #-}
   sizeOf _ = 32
+  {-# INLINE sizeOf #-}
   peek = evalStruct $ TSNode <$> peekStruct
                              <*> peekStruct
                              <*> peekStruct
                              <*> peekStruct
                              <*> peekStruct
-                             <*> peekStruct
-  poke ptr (TSNode o1 o2 o3 o4 p1 p2) = flip evalStruct ptr $ do
-    pokeStruct o1
-    pokeStruct o2
-    pokeStruct o3
+  {-# INLINE peek #-}
+  poke ptr (TSNode sb sp o4 p1 p2) = flip evalStruct ptr $ do
+    pokeStruct sb
+    pokeStruct sp
     pokeStruct o4
     pokeStruct p1
     pokeStruct p2
+  {-# INLINE poke #-}
 
 instance Functor Struct where
-  fmap f a = Struct (\ p -> do
-    (a', p') <- runStruct a p
-    let fa = f a'
-    fa `seq` p' `seq` pure (fa, castPtr p'))
+  fmap f a = Struct go where
+    go p = do
+      (a', p') <- runStruct a p
+      let fa = f a'
+      fa `seq` p' `seq` pure (fa, castPtr p')
```

Comment:
Technically these `seq`s violate the `Monad` laws, but I think that’s perfectly cromulent, given that this is used in this one place and needs to go as fast as humanly possible.